### PR TITLE
Less io less nt

### PR DESCRIPTION
**Pull request by j042**
Remove recursive basis nature and IOs that we don't use. A lot of tests become irrelevant and the functionalities we use are tested in splinepy.

### CODE REVIEW

File: Sources/BSplineLib/ParameterSpaces/knot_vector.cpp [hunk 1]
```diff
@@ -81,15 +81,37 @@ Knot const& KnotVector::operator[](Index const& index) const {
...
 int KnotVector::GetSize() const { return knots_.size(); }
 
 Knot const& KnotVector::GetFront() const { return knots_[0]; }
 
 Knot const& KnotVector::GetBack() const { return knots_[knots_.size() - 1]; }
 
+void KnotVector::UpdateKnot(const int id, Knot const& knot) {
+  if (knots_[id - 1] > knot || knots_[id + 1] < knot) {
+    throw RuntimeError(
+        "KnotVector::UpdateKnot - updating not must be non-decreasing.");
```
Comment (jzwar) [inline Sources/BSplineLib/ParameterSpaces/knot_vector.cpp]:
```suggestion
        "KnotVector::UpdateKnot - updated knot must be non-decreasing.");
```

File: Sources/BSplineLib/ParameterSpaces/parameter_space.hpp [hunk 2]
```diff
@@ -156,6 +154,24 @@ RecursiveCombine(const Array<Vector<ValueType>, array_dim>& factors,
                                    result);
 }
 
+template<typename T>
+struct TemporaryArray {
```
Comment (jzwar) [inline Sources/BSplineLib/ParameterSpaces/parameter_space.hpp:158]: Did you just reimplement a minimalist version of std::array?

File: Sources/BSplineLib/VectorSpaces/weighted_vector_space.inc [hunk 3]
```diff
@@ -80,26 +80,14 @@ WeightedVectorSpace<dimensionality>::
...
-                           tolerance));
-    minimum_weight = std::min(
-        Weight{homogeneous_coordinate[dimensionality].Get()},
-        minimum_weight,
-        std::bind(
-            utilities::IsLess<typename Weight::Name_, typename Weight::Type_>,
-            _1,
-            _2,
-            tolerance));
+    maximum_distance = std::max(utilities::std_container_operations::TwoNorm(
```
Comment (jzwar) [inline Sources/BSplineLib/VectorSpaces/weighted_vector_space.inc:86]: TBH I don't understand what you are doing here...

File: Sources/BSplineLib/ParameterSpaces/parameter_space.hpp [hunk 4]
```diff
@@ -259,15 +267,14 @@ class ParameterSpace {
...
       Tolerance const& tolerance = kEpsilon) const;
+
+  virtual KnotSpans_
+  FindKnotSpans(ParametricCoordinate_ const& parametric_coordinate,
+                Tolerance const& tolerance = kEpsilon) const;
   virtual BezierInformation_
   DetermineBezierExtractionKnots(Dimension const& dimension,
                                  Tolerance const& tolerance = kEpsilon) const;
 
-  virtual Type_
```
Comment (jzwar) [inline Sources/BSplineLib/ParameterSpaces/parameter_space.hpp:266]: Please add a comment to the EvaluateBasisFunctionPerDimension routine, so it is clear from where you took the algorithm (with citation, page and alg. ref). Also, the class documentation seems outdated

**Review by jzwar — APPROVED** [review]
Thank you for removing the recursive nature of BSplines. Although it makes the parametric space a bit more intuitive, the complicated building process of the tree makes it too hard to maintain. The changes look good to me, please update some parts of the documentation. If the array is really necessary, I am fine with it, though I would prefer you rely more on std implementations.

File: Sources/BSplineLib/ParameterSpaces/knot_vector.cpp (hunk 1 above)
Comment (j042) [reply Sources/BSplineLib/ParameterSpaces/knot_vector.cpp]: good catch

File: Sources/BSplineLib/ParameterSpaces/parameter_space.hpp (hunk 2 above)
Comment (j042) [reply Sources/BSplineLib/ParameterSpaces/parameter_space.hpp:158]: I guess it's rather close to std::vector

File: Sources/BSplineLib/VectorSpaces/weighted_vector_space.inc (hunk 3 above)
Comment (j042) [reply Sources/BSplineLib/VectorSpaces/weighted_vector_space.inc:86]: Originally I wanted to just remove this, but it seems like it is used in remove_knots / reduce_degree to measure some relative tolerance

File: Sources/BSplineLib/ParameterSpaces/parameter_space.hpp (hunk 4 above)
Comment (j042) [reply Sources/BSplineLib/ParameterSpaces/parameter_space.hpp:266]: done

**Pull request merged**